swarm/storage/request.go

```diff
@@ -0,0 +1,57 @@
+// Copyright 2018 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package storage
+
+import (
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/p2p/enode"
+	"github.com/ethereum/go-ethereum/swarm/network/timeouts"
+)
+
+type Request struct {
+	Addr        Address  // chunk address
+	Origin      enode.ID // who is sending us that request? we compare Origin to the suggested peer from RequestFromPeers
+	PeersToSkip sync.Map // peers not to request chunk from
+	HopCount    uint8    // number of forwarded requests (hops)
+}
+
+// NewRequest returns a new instance of Request based on chunk address skip check and
+// a map of peers to skip.
+func NewRequest(addr Address, hopCount uint8) *Request {
+	return &Request{
+		Addr:        addr,
+		HopCount:    hopCount,
+		PeersToSkip: sync.Map{},
+	}
+}
+
+// SkipPeer returns if the peer with nodeID should not be requested to deliver a chunk.
+// Peers to skip are kept per Request and for a time period of FailedPeerSkipDelay.
+func (r *Request) SkipPeer(nodeID string) bool {
+	val, ok := r.PeersToSkip.Load(nodeID)
+	if !ok {
+		return false
+	}
+	t, ok := val.(time.Time)
+	if ok && time.Now().After(t.Add(timeouts.FailedPeerSkipDelay)) {
```

Comment:
You will have to elaborate on what you mean. This definitely happens - basically, if we try to fetch a chunk from a peer and it doesn't respond within the `SearchTimeout`, we ignore this peer for the next `FailedPeerSkipDelay`.

It is argueable whether you want to try again a peer that has failed you once - in tests this happens if you request a chunk that is not synced yet, and it gets synced only after the first request. Then the peer is added to the `PeersToSkip` list, and on second try - it returns the chunk.

AFAIK we agree that we want Swarm to keep trying to find a chunk until a given timeout is reached - and the user should be able to specify the timeout (it could be very long, for example 1 minute). Then the retry mechanism is basically part of Swarm.

What you suggest is removing this, and having the retry mechanism outside of Swarm - something that I believe we don't want. If we remove this, then we can handle requests very quickly and just return 404 to the user, and he is fine with trying again later...

Reply:
so the retry mechanism is something i think we should properly define first. we should define a search heuristic that should be (dis)proven based on empiric results. a request refers to a chunk address which is being requested. 

i essentially claim that on a stable swarm where peers don't heavily churn - i don't see any reason why a peer's ability to deliver a certain chunk should change within the scope of a request life-cycle.